plugins/out_es/es_conf_parse.c

```diff
@@ -0,0 +1,338 @@
+#include <fluent-bit/flb_output_plugin.h>
+#include <fluent-bit/flb_aws_credentials.h>
+
+#include "es_conf_parse.h"
+
+int flb_es_conf_set_cloud_credentials(const char *cloud_auth,
+                                      struct flb_elasticsearch_config *ec)
+{
+    /* extract strings */
+    int items = 0;
+    struct mk_list *toks;
+    struct mk_list *head;
+    struct flb_split_entry *entry;
+
+    if (!cloud_auth) {
+        return 0;
+    }
+
+    toks = flb_utils_split((const char *)cloud_auth, ':', -1);
+    mk_list_foreach(head, toks) {
+        items++;
+        entry = mk_list_entry(head, struct flb_split_entry, _head);
+        if (items == 1) {
+            ec->cloud_user = flb_strdup(entry->value);
+            ec->own_cloud_user = FLB_TRUE;
+        }
+        if (items == 2) {
+            ec->cloud_passwd = flb_strdup(entry->value);
+            ec->own_cloud_passwd = FLB_TRUE;
+        }
+    }
+    flb_utils_split_free(toks);
+
+    return 0;
+}
+
+/*
+ * extract_cloud_host extracts the public hostname
+ * of a deployment from a Cloud ID string.
+ *
+ * The Cloud ID string has the format "<deployment_name>:<base64_info>".
+ * Once decoded, the "base64_info" string has the format "<deployment_region>$<elasticsearch_hostname>$<kibana_hostname>"
+ * and the function returns "<elasticsearch_hostname>.<deployment_region>" token.
+ */
+static flb_sds_t extract_cloud_host(const char *cloud_id, struct flb_elasticsearch *ctx)
+{
+
+    char *colon;
+    char *region;
+    char *host;
+    char *port = NULL;
+    char buf[256] = {0};
+    char cloud_host_buf[256] = {0};
+    const char dollar[2] = "$";
+    size_t len;
+    int ret;
+
+    /* keep only part after first ":" */
+    colon = strchr(cloud_id, ':');
+    if (colon == NULL) {
+        return NULL;
+    }
+    colon++;
+
+    /* decode base64 */
+    ret = flb_base64_decode((unsigned char *)buf, sizeof(buf), &len,
+                            (unsigned char *)colon, strlen(colon));
+    if (ret) {
+        flb_plg_error(ctx->ins, "cannot decode cloud_id");
+        return NULL;
+    }
+    region = strtok(buf, dollar);
+    if (region == NULL) {
+        return NULL;
+    }
+    host = strtok(NULL, dollar);
+    if (host == NULL) {
+        return NULL;
+    }
+
+    /*
+     * Some cloud id format is "<deployment_region>$<elasticsearch_hostname>:<port>$<kibana_hostname>" .
+     *   e.g. https://github.com/elastic/beats/blob/v8.4.1/libbeat/cloudid/cloudid_test.go#L60
+     *
+     * It means the variable "host" can contains ':' and port number.
+     */
+    colon = strchr(host, ':');
+    if (colon != NULL) {
+        /* host contains host number */
+        *colon = '\0'; /* remove port number from host */
+        port = colon+1;
+    }
+
+    strcpy(cloud_host_buf, host);
+    strcat(cloud_host_buf, ".");
+    strcat(cloud_host_buf, region);
+    if (port != NULL) {
+        strcat(cloud_host_buf, ":");
+        strcat(cloud_host_buf, port);
+    }
```

Comment:
We need to use strict length version of strncat and strcpy at least.
Or, Fluent Bit offers the flb_sds function APIs which can handling more safely to process strings.
Could you use them if you could do?

Reply:
This part is original code - refer to https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es_conf.c#L92-L95 - moved to a new location. More changes we introduce into this pull request - harder it will be accept / test this pull request. I'm not against of fixing issues of original code in this pull request too (in dedicated commit(s)), but I'd like to ensure it is really needed and requesters of fixes realize the source of these issues.